database/ticket.go

```diff
@@ -244,6 +291,84 @@ func (vdb *VspDatabase) UpdateTicket(ticket Ticket) error {
 	})
 }
 
+// AltSigHistory retrieves the history of changes to a ticket's alternate signature.
+func (vdb *VspDatabase) AltSigHistory(ticketHash string) ([]*AltSigHistory, error) {
+	vdb.ticketsMtx.RLock()
+	defer vdb.ticketsMtx.RUnlock()
+
+	history := make([]*AltSigHistory, 0)
+	return history, vdb.db.View(func(tx *bolt.Tx) error {
+		ticketBkt := tx.Bucket(vspBktK).Bucket(ticketBktK).Bucket([]byte(ticketHash))
+		if ticketBkt == nil {
+			return nil
+		}
+		histBkt := ticketBkt.Bucket(altSigHistoryK)
+		if histBkt == nil {
+			return nil
+		}
+		// BucketN includes the history bucket.
+		nBkts := histBkt.Stats().BucketN - 1
+		for i := 0; i < nBkts; i++ {
+			bkt := histBkt.Bucket([]byte{byte(i)})
+			h := new(AltSigHistory)
+			h.Addr = string(bkt.Get(altSigHistoryAddrK))
+			h.Req = bkt.Get(altSigHistoryReqK)
+			h.Sig = bkt.Get(altSigHistorySigK)
+			history = append(history, h)
+		}
+		return nil
+	})
+}
+
+// AddAltSigHistory adds an alternate signature record to history. If no bucket
+// exists for this ticket an error is returned. If the altSigHistory bucket does
+// not yet exist for the ticket, one is created. Records are keyed by number of
+// records plus one. Only maxAltSigs are allowed, and trying to add more returns
+// error. These records should never be deleted or altered as the server needs
+// them to prove the client's current alternate signature was indeed chosen by
+// the client.
+func (vdb *VspDatabase) AddAltSigHistory(ticketHash string, history *AltSigHistory) error {
+	vdb.ticketsMtx.Lock()
+	defer vdb.ticketsMtx.Unlock()
+	return vdb.db.Update(func(tx *bolt.Tx) error {
+		ticketBkt := tx.Bucket(vspBktK).Bucket(ticketBktK).Bucket([]byte(ticketHash))
+		if ticketBkt == nil {
+			return fmt.Errorf("not bucket for ticket %v", ticketHash)
```

Comment:
s / not bucket for ticket / no bucket for ticket.

Reply:
ok